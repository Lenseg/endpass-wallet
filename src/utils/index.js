export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());
