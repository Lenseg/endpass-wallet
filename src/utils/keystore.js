import keythereum from 'keythereum';
import bs58check from 'bs58check';
import { kdfParams } from '@/config';

// Monkey patch keythereum to skip generating address for private keys
// This allows us to encrypt private keys of arbitrary length, and
// conforms better to the Ethereum keystore V3 spec, which does not include
// the address for privacy reasons.
// See https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition#alterations-from-version-1
keythereum.privateKeyToAddress = function(pk) {
  return '';
};

export default {
  // Encrypts a private key Buffer into a V3 keystore object
  // The exported keystore does NOT include an address
  encrypt(password, privateKey) {
    // Generate random salt and iv for each encryption
    let dk = keythereum.create();
    let options = {
      kdf: kdfParams.kdf,
      kdfparams: kdfParams,
    };
    let encrypted = keythereum.dump(
      password,
      privateKey,
      dk.salt,
      dk.iv,
      options,
    );
    delete encrypted.address;
    return encrypted;
  },

  // Decrypts a V3 keystore object into a private key Buffer
  decrypt(password, keystore) {
    return keythereum.recover(password, keystore);
  },

  // Encode to Base58Check
  encodeBase58(key) {
    return bs58check.encode(key);
  },

  // Decode from Base58Check
  decodeBase58(key) {
    return bs58check.decode(key);
  },
};
