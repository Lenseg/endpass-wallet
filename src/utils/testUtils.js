const toKebab = string =>
  string.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let dashName;

    if (stubComponent.name) {
      dashName = toKebab(stubComponent.name);
    }

    stubs[dashName] = {
      render(createElement) {
        return createElement(dashName, this.$slots.default);
      },
    };

    return stubs;
  }, {});
}
