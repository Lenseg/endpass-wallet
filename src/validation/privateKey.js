import ethUtils from 'ethereumjs-util';
export default {
  getMessage(field, params, data) {
    return (data && data.message) || 'Something went wrong';
  },
  validate(value, args) {
    let isKey;
    try {
      isKey = ethUtils.isValidPrivate(
        Buffer.from(value.replace(/^0x/, ''), 'hex'),
      );
    } catch (e) {
      return {
        valid: false,
        data: { message: e.message },
      };
    }
    return {
      valid: isKey,
      data: !isKey ? { message: 'This is not a valid private key' } : undefined,
    };
  },
};
