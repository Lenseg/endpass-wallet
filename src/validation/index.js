import Vue from 'vue';
import { default as VeeValidate, Validator } from 'vee-validate';
import address from './address';
import privateKey from './privateKey';
import publicKey from './publicKey';
import seedPhrase from './seedPhrase';
import hex from './hex';

Validator.extend('address', address);
Validator.extend('private_key', privateKey);
Validator.extend('public_key', publicKey);
Validator.extend('seed_phrase', seedPhrase);
Validator.extend('hex', hex);

Vue.use(VeeValidate);
