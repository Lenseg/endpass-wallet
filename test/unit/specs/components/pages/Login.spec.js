import { shallow, createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import Notifications from 'vue-notification';
import validation from '@/validation';

import LoginPage from '@/components/pages/Login.vue';
import { generateStubs } from '@/utils/testUtils';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Notifications);

describe('LoginPage', () => {
  let wrapper;
  let actions;
  let store;
  let options;

  describe('render', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = shallow(LoginPage, {
        stubs: generateStubs(LoginPage)
      });
    });

    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('login-page');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    describe('the send button', () => {
      it('should change "disable" attribute', () => {
        wrapper.setData({ termsAccepted: true });
        expect(wrapper.find('v-button').attributes().disabled).toBeFalsy();

        wrapper.setData({ termsAccepted: false });
        expect(wrapper.find('v-button').attributes().disabled).toBeTruthy();
      });
    });
  });

  beforeEach(() => {
    actions = {
      login: jest.fn(() => Promise.resolve()),
    };

    store = new Vuex.Store({
      modules: {
        accounts: {
          namespaced: true,
          actions,
        },
      },
    });

    options = {
      store,
      localVue,
    };

    wrapper = shallow(LoginPage, options);
  });

  it('should call login action when click', async () => {
    wrapper = mount(LoginPage, options);

    const email = '123@123.com';

    wrapper.setData({ email });

    wrapper.find('a#send-button').trigger('click');

    await new Promise(res => setTimeout(res, 50));
    expect(actions.login).toHaveBeenCalledTimes(1);
    expect(actions.login).toBeCalledWith(expect.any(Object), email, undefined);

    await actions.login();
    await wrapper.vm.$nextTick();

    expect(wrapper.find('p#success-message')).toBeTruthy();
  });

  // it('should validate data', async () => {
  //   wrapper = mount(LoginPage, {
  //     store,
  //     localVue,
  //   });

  //   const { errors } = wrapper.vm;

  //   wrapper.setData({
  //     email: '',
  //   });

  //   await wrapper.vm.$validator.validateAll();

  //   expect(errors.first('email').includes('required')).toBeTruthy();

  //   wrapper.setData({
  //     email: '123@123.com',
  //   });

  //   await wrapper.vm.$validator.validateAll();

  //   expect(errors.has('email')).toBeFalsy();

  //   wrapper.setData({
  //     email: '123',
  //   });

  //   await wrapper.vm.$validator.validateAll();

  //   expect(
  //     errors.first('email').includes('must be a valid email')
  //   ).toBeTruthy();
  // });
});
