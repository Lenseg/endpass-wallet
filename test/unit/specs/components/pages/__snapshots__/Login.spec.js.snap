// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage render should render component 1`] = `
<div
  class="app-page send-page"
>
  <div
    class="section"
  >
    <div
      class="container is-narrow"
    >
      <div
        class="card app-card main-app-card"
      >
        <div
          class="card-header"
        >
          <h1
            class="card-header-title"
          >
            Login/Register
          </h1>
        </div>
         
        <div
          class="card-content"
        >
          <v-form
            id="login"
          >
            <v-input
              aria-describedby="email"
              id="email"
              label="Email"
              name="email"
              placeholder="Your email"
              required=""
              validator="required|email"
            />
             
            <v-checkbox>
              
              I accept 
              <a
                href="https://endpass.com/terms/"
                target="_blank"
              >
                Terms of Service
              </a>
              
              and 
              <a
                href="https://endpass.com/privacy/"
                target="_blank"
              >
                Privacy Policy
              </a>
            </v-checkbox>
             
            <v-button
              classname="is-primary is-medium"
              disabled="true"
              id="send-button"
            >
              Send
            </v-button>
          </v-form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
